modules/cudaoptflow/src/cuda/pyrlk.cu

```diff
@@ -60,53 +62,325 @@ namespace pyrlk
+
+    texture<ushort, cudaTextureType2D, cudaReadModeNormalizedFloat> tex_I16U(false, cudaFilterModeLinear, cudaAddressModeClamp);
+    texture<ushort4, cudaTextureType2D, cudaReadModeNormalizedFloat> tex_I16UC4(false, cudaFilterModeLinear, cudaAddressModeClamp);
+
+    texture<int, cudaTextureType2D, cudaReadModeElementType> tex_I32S(false, cudaFilterModePoint, cudaAddressModeClamp);
+    texture<int4, cudaTextureType2D, cudaReadModeElementType> tex_I32SC4(false, cudaFilterModePoint, cudaAddressModeClamp);
+
     texture<float, cudaTextureType2D, cudaReadModeElementType> tex_If(false, cudaFilterModeLinear, cudaAddressModeClamp);
     texture<float4, cudaTextureType2D, cudaReadModeElementType> tex_If4(false, cudaFilterModeLinear, cudaAddressModeClamp);
+
     texture<uchar, cudaTextureType2D, cudaReadModeElementType> tex_Ib(false, cudaFilterModePoint, cudaAddressModeClamp);
 
+    texture<uchar, cudaTextureType2D, cudaReadModeNormalizedFloat> tex_J8U(false, cudaFilterModeLinear, cudaAddressModeClamp);
+    texture<uchar4, cudaTextureType2D, cudaReadModeNormalizedFloat> tex_J8UC4(false, cudaFilterModeLinear, cudaAddressModeClamp);
+
+    texture<ushort, cudaTextureType2D, cudaReadModeNormalizedFloat> tex_J16U(false, cudaFilterModeLinear, cudaAddressModeClamp);
+    texture<ushort4, cudaTextureType2D, cudaReadModeNormalizedFloat> tex_J16UC4(false, cudaFilterModeLinear, cudaAddressModeClamp);
+
+    texture<int, cudaTextureType2D, cudaReadModeElementType> tex_J32S(false, cudaFilterModePoint, cudaAddressModeClamp);
+    texture<int4, cudaTextureType2D, cudaReadModeElementType> tex_J32SC4(false, cudaFilterModePoint, cudaAddressModeClamp);
+
     texture<float, cudaTextureType2D, cudaReadModeElementType> tex_Jf(false, cudaFilterModeLinear, cudaAddressModeClamp);
     texture<float4, cudaTextureType2D, cudaReadModeElementType> tex_Jf4(false, cudaFilterModeLinear, cudaAddressModeClamp);
 
-    template <int cn> struct Tex_I;
-    template <> struct Tex_I<1>
+
+    template <int cn, typename T = float> struct Tex_I{};
+
+    template <> struct Tex_I<1, uchar>
+    {
+        static __device__ __forceinline__ float read(float x, float y)
+        {
+            return tex2D(tex_I8U, x, y);
+        }
+        static __host__ __forceinline__ void bindTexture_(PtrStepSz<uchar>& I)
+        {
+            bindTexture(&tex_I8U, I);
+        }
+    };
+    template <> struct Tex_I<1, ushort>
+    {
+        static __device__ __forceinline__ float read(float x, float y)
+        {
+            return tex2D(tex_I16U, x, y);
+        }
+        static __host__ __forceinline__ void bindTexture_(PtrStepSz<ushort>& I)
+        {
+            bindTexture(&tex_I16U, I);
+        }
+    };
+    template <> struct Tex_I<1, int>
+    {
+        static __device__ __forceinline__ float read(float x, float y)
+        {
+            return tex2D(tex_I32S, x, y);
```

Comment:
Shouldn't you convert the value to `[0, 1]` range to match `cudaReadModeNormalizedFloat` behavior?